src/app/records/PresentationNodeLeaf.tsx

```diff
@@ -62,3 +70,65 @@ export default function PresentationNodeLeaf({
     </>
   );
 }
+
+/**
+ * Sorts the passed node if it has a DimRecord[] | DimCollectible[] | DimMetric[]
+ */
+function sortRecords(node: DimPresentationNodeLeaf): DimPresentationNodeLeaf {
+  const sortedNode = { ...node };
+  // RECORDS
+  if (sortedNode.records) {
+    sortedNode.records = sortBy(sortedNode.records, (record) => {
+      //  Triumph is already completed so move it to back of list.
+      if (
+        record.recordComponent.state & DestinyRecordState.RecordRedeemed ||
+        record.recordComponent.state & DestinyRecordState.CanEquipTitle ||
+        !record.recordComponent.state
+      ) {
+        return -1;
+      }
+
+      // check which key is used to track progress
+      let objectives;
+      if (record.recordComponent.intervalObjectives) {
+        objectives = record.recordComponent.intervalObjectives;
+      } else if (record.recordComponent.objectives) {
+        objectives = record.recordComponent.objectives;
+      } else {
+        // its a legacy triumph so it has no objectives and is not completed
+        return 0;
+      }
+
+      //  Sum up the progress
+      let totalProgress = 0;
+      for (const x of objectives) {
+        // Some progress bars exceed its completionValue
+        if (x.complete) {
+          totalProgress += 1;
+          continue;
+        }
+        totalProgress += x.progress! / x.completionValue;
+      }
+      return totalProgress / objectives.length;
+    }).reverse();
+    // COLLECTIBLES
+  } else if (sortedNode.collectibles) {
+    //  Only need to push owned items to the back of the list
+    sortedNode.collectibles = sortBy(sortedNode.collectibles, (collectible) => {
+      if (collectible.state & DestinyCollectibleState.NotAcquired) {
+        return -1;
+      }
+      return 0;
+    });
```

Comment:
https://github.com/DestinyItemManager/DIM/pull/9602#discussion_r1264278366

Especially once you're no longer modifying things in place, each of these sorts can be in its own function.

Reply:
So is this how you wanted the functions to be structured? If so how should it be used by the components?
![image](https://github.com/DestinyItemManager/DIM/assets/99226958/c9cc4805-ddc7-43a3-a442-8bd6c7980a39)

Reply:
Right - one function per type of thing. Then you don't need `sortNode` since you're going to sort them individually anyway.

Reply:
And have the components use it like this?
![image](https://github.com/DestinyItemManager/DIM/assets/99226958/1fa50e6b-3350-4c66-b506-e018c7578ce5)

Reply:
Not quite. Try something like this:

```ts
{(sortRecordProgression ? sortCollectibles(node.collectibles) : node.collectibles).map((collectible) => )
```

BTW you can paste code into GitHub comments, you don't need to screenshot!

Reply:
everything should be good now